apps/src/lib/util/AnalyticsReporter.js

```diff
@@ -0,0 +1,134 @@
+import {
+  init,
+  track,
+  Identify,
+  identify
+  // setUserId
+} from '@amplitude/analytics-browser';
+import {currentLocation} from '@cdo/apps/utils';
+
+// A flag that can be toggled to send events regardless of environment
+const ALWAYS_PUT = false;
+
+const Environments = {
+  production: 'production',
+  levelbuilder: 'levelbuilder',
+  test: 'test',
+  staging: 'staging',
+  adhoc: 'adhoc',
+  development: 'development',
+  unknown: 'unknown'
+};
+
+class AnalyticsReporter {
+  constructor() {
+    // Get the API key from the DOM. See application.html.haml layout.
+    const element = document.querySelector('script[data-amplitude-api-key]');
+    const api_key = element ? element.dataset.amplitudeApiKey : '';
+    // Let Amplitude handle missing/invalid keys.
+    if (this.shouldPutRecord(ALWAYS_PUT)) {
+      init(api_key);
+    }
+  }
+
+  setUserProperties(userId, userType, signInState) {
+    const identifyObj = new Identify();
+    const formattedUserId = this.formatUserId(userId);
+    // Dont set user id pending privacy conversations
+    // setUserId(formattedUserId);
+    identifyObj.set('userType', userType);
+    identifyObj.set('signInState', signInState);
+
+    if (!this.shouldPutRecord(ALWAYS_PUT)) {
+      this.log(
+        `User properties: userId: ${formattedUserId}, userType: ${userType}, signInState: ${signInState}`
+      );
+    }
+    identify(identifyObj);
+  }
+
+  sendEvent(eventName, payload) {
+    if (this.shouldPutRecord(ALWAYS_PUT)) {
+      track(eventName, payload);
+    } else {
+      this.log(`${eventName}. Payload: ${JSON.stringify({payload})}`);
+    }
+  }
+
+  log(message) {
+    console.log(`[AMPLITUDE ANALYTICS EVENT]: ${message}`);
+  }
+
+  formatUserId(userId) {
+    const userIdString = userId.toString() || 'none';
+    if (!userId) {
+      return userIdString;
+    }
+    if (this.isProductionEnvironment()) {
+      return userIdString.padStart(5, '0');
+    } else {
+      const environment = this.getEnvironment();
+      return `${environment}-${userIdString}`;
+    }
+  }
+
+  /**
+   * Returns whether the request should be sent through to AWS Firehose.
+   * @param {boolean} alwaysPut An override to default environment behavior.
+   * @return {boolean} Whether the request should be sent through to AWS
+   *   Firehose.
+   */
+  shouldPutRecord(alwaysPut) {
+    if (alwaysPut) {
+      return true;
+    }
+    if (this.isTestEnvironment() || this.isDevelopmentEnvironment()) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Returns the current environment.
+   * @return {string} The current environment, e.g., "staging" or "production".
+   */
+  getEnvironment() {
+    const hostname = currentLocation().hostname;
+    if (hostname.includes('adhoc')) {
+      // As adhoc hostnames may include other keywords, check it first.
+      return Environments.adhoc;
```

Comment:
Ah yes haha I spoke so soon. I'm diggin the detailed environment list. I do remember that events on the Amplitude side can only be shared across two API keys, so I wonder what the solution will be there.

Reply:
We just chatted about this and we'll have two projects: one for production and one for staging. If for some reason we need to send events from a non-staging non-production environment, we will send those events to the staging project. We won't send any non-production events to the production project.

Reply:
Gotcha! I gave this PR a formal ✅, assuming there might be some updates to the environment section below given the above^.

Reply:
The environment section is just a list so I wasn't going to make any changes in case we need some logic around the other environments later. `shouldPutRecord` contains the logic for which environments to send events from (which I might update as a followup to `shouldSendRecord` for consistency).